### PR TITLE
Fixes and improvements to core/codegen

### PR DESCRIPTION
**Pull request by ben-clayton**

### CODE REVIEW

File: core/codegen/builder.go [hunk 1]
```diff
@@ -66,7 +66,51 @@ func (b *Builder) CallIndirect(f *Value, args ...*Value) *Value {
...
+	if sig.Result == b.m.Types.Void {
+		name = ""
+	}
+	return b.val(sig.Result, b.llvm.CreateCall(fn, l, name))
+}
+
+// Invoke invokes the function f with the specified arguments.
+// If an exception is thrown while calling f, then cleanup will be called before
+// rethrowing the exception.
+func (b *Builder) Invoke(f *Function, cleanup func(), args ...*Value) *Value {
```
Comment (AWoloszyn) [inline core/codegen/builder.go]: Are we using exceptions here to just plumb through errors? 
I.e. so we don't have to return true/false from every method (slice access etc).

**Comment by AWoloszyn** [comment]
It looks like the MacOS bug is a flake. Unrelated to this MR.

File: core/codegen/builder.go (hunk 1 above)
Comment (ben-clayton) [reply core/codegen/builder.go]: Yes - these are for `abort`s and `assert`s. The exceptions here are 'zero cost', and increse performance by about 10% - 15% based on my measurements.

**Comment by ben-clayton** [comment]
Thank you.

**Pull request merged**